### PR TITLE
Extractor functions

### PR DESCRIPTION
**Pull request by nystrom**
This PR implements extractor functions for Rematch.

Patterns can use _extractor functions_ (also known as _active patterns_). These are just any function that takes a value to match and returns either `nothing` (indicating match failure) or a tuple that decomposes the value. The tuple is then matched against other patterns. 

An extractor function must have a lowercase name to distinguish it from a struct name. [I'd like to relax this requirement, if possible, but we should discuss.]

An extractor function must take one argument--the value to be matched against--and should return either one value (for nullary and unary patterns), or a tuple of values (for 2+-ary patterns). Returning `nothing` indicates the extractor does not match. 

For example to destruct an array into its head and tail:

```julia
function cons(xs)
    if isempty(xs)
        nothing
    else
        ([xs[1], xs[2:end]])
    end
end

@match [1,2,3] begin
    cons(x, xs) => @assert x == 1 && xs == [2,3]
end
```

The main code changes are in `handle_destruct` for the `T_(subpatterns__)` case.

### CODE REVIEW

**Comment by cscherrer** [comment]
This looks really nice @nystrom . Am I understanding right that the semantics are that

```julia
@match foo begin
    cons(x, xs) => bar
end
```

is equivalent to

```julia
@match cons(foo) begin
    (x, xs) => bar
end
```

so it seems the function is converted to its inverse? I haven't used active patterns much, is this a common way to write things?

**Comment by nystrom** [comment]
Those two cases are equivalent. Yes, the extractor is basically the inverse of the constructor.
But you can't just all the inverse function directly on the scrutinee after `@match` because you need to call different functions in different case arms. For example:

```julia
@match foo begin
   cons(x, cons(y, ys)) => foo
   cons(x, xs) => bar
   [] => baz
end
```
calls `cons` on `foo` in both the first and second patterns, but not in the third.

File: src/Rematch.jl [hunk 1]
```diff
@@ -128,42 +128,76 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
...
-                if typeof($(esc(T))) <: Function
-                    throw(LoadError("Attempted to match on a function", @__LINE__, AssertionError("Incorrect match usage")))
+        # struct or extractor call
+        if !isempty(string(T)) && islowercase(first(string(T)))
+            # Extractor call.
+            result = gensym("unapply")
+            len = length(subpatterns)
+            # Extractor call.
+            # The function should take one argument and return either `nothing`, if the argument does not match,
+            # or a tuple if it does match.
```
Comment (comnik) [inline src/Rematch.jl:138]: Please make sure to also break comments at 92chars to stay consistent with the style guide.

File: test/rematch.jl [hunk 2]
```diff
@@ -47,6 +47,190 @@ assertion_error = (VERSION >= v"0.7.0-DEV") ? LoadError : AssertionError
...
+        @test x == nothing
+    end
+
+    function cons(xs)
+        if isempty(xs)
+            nothing
+        else
+            (xs[1], xs[2:end])
+        end
+    end
```
Comment (comnik) [inline test/rematch.jl:68]: Please use [explicit returns for all multiline functions](https://github.com/relationalai/raistyle#method-definitions). I know its minor but we should be getting used to sticking to the style guide in our OSS as well :)

File: test/rematch.jl [hunk 3]
```diff
@@ -47,6 +47,190 @@ assertion_error = (VERSION >= v"0.7.0-DEV") ? LoadError : AssertionError
...
+
+        blk(a, y, b)
+    end
+
+    function balance(t::T)
+        @match t begin
+            ( blk(red(red(a, x, b), y, c), z, d) ||
+              blk(red(a, x, red(b, y, c)), z, d) ||
+              blk(a, x, red(red(b, y, c), z, d)) ||
+              blk(a, x, red(b, y, red(c, z, d))) ) => red(blk(a, x, b), y, blk(c, z, d))
```
Comment (comnik) [inline test/rematch.jl:205]: Well this is kinda cool 😄

File: src/Rematch.jl [hunk 4]
```diff
@@ -128,42 +128,76 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
...
         else
-            # Pattern uses named arguments to refer to fields e.g. Foo(x=0, y=true)
-            # Could assert that the expected field names are a subset of the actual field names.
-            # Can omit the assertion because if the field doesn't exist getfield() will fail with "type $T has no field $field".
-        end
-        quote
-            # I would prefer typeof($value) == $(esc(T)) but this doesn't convey type information in Julia 0.6
-            $value isa $(esc(T)) &&
-            $(handle_destruct_fields(value, pattern, subpatterns, length(subpatterns), :getfield, bound, asserts; allow_splat=false))
+            # Struct.
```
Comment (comnik) [inline src/Rematch.jl:164]: The diff got chopped up in a really annoying way. I think you didn't change anything about the existing implementation for matching structs, right?

File: src/Rematch.jl [hunk 5]
```diff
@@ -128,42 +128,76 @@ function handle_destruct(value::Symbol, pattern, bound::Set{Symbol}, asserts::Ve
...
                 end
-                # This assertion is necessary:
-                # If $expected_fieldcount < $actual_fieldcount, to catch missing fields.
-                # If $expected_fieldcount > $actual_fieldcount, to avoid a BoundsError.
-                $expected_fieldcount = evaluated_fieldcount($(esc(T)))
-                $actual_fieldcount = $(esc(len))
-                if $expected_fieldcount != $actual_fieldcount
-                    error("Pattern field count is $($actual_fieldcount) expected $($expected_fieldcount)")
+            else
+                # If there is more than one subpattern, the result value is matched against a tuple pattern.
```
Comment (comnik) [inline src/Rematch.jl:153]: Having looked at the rematch source for the first time: what is a tuple pattern? Maybe you could take this opportunity to add docstrings / comments to `handle_destruct_fields` and `handle_destruct` :)

File: test/rematch.jl (hunk 2 above)
Comment (nystrom) [reply test/rematch.jl:68]: Fixed. I also went through the entire file to fix the line lengths.

File: src/Rematch.jl (hunk 4 above)
Comment (nystrom) [reply src/Rematch.jl:164]: no, just whitespace.

File: src/Rematch.jl (hunk 5 above)
Comment (nystrom) [reply src/Rematch.jl]: docstrings added

File: src/Rematch.jl (hunk 1 above)
Comment (nystrom) [reply src/Rematch.jl]: Line lengths fixed throughout the file.

**Comment by nystrom** [comment]
I was thinking that rather than using uppercase/lowercase to distinguish between structs and extractors, we could just prepend an operator like `~`.

```julia
@match foo begin
   ~Foo(x,y,z) => ... # call extractor Foo
   Foo(x,y,z) => ... # match struct Foo
end
```

I'm converting this back to a draft PR.

File: src/Rematch.jl (hunk 4 above)
Comment (rai-nhdaly) [reply src/Rematch.jl]: @comnik you can disable whitespace in github's diff, i dunno if you've seen that? :)

File: test/rematch.jl [hunk 6]
```diff
@@ -47,6 +47,190 @@ assertion_error = (VERSION >= v"0.7.0-DEV") ? LoadError : AssertionError
...
 end
 
+@testset "Match using extractors" begin
+    function sub1(x) x+1 end
+
+    # sub1(4) == 3
+    let x = nothing
+        @test (@match 3 begin
+           sub1(x) => x
+        end) == 4
```
Comment (rai-nhdaly) [inline test/rematch.jl]: I don't understand why this returns `4`. I could understand if it returned `2` or `3`, but i don't quite understand how it returns `4`.

The definition of `sub1(x)` is `sub1(x) = x+1`. So I take `@match 3 begin   sub1(x) => x  end` to mean "if `3` matches `sub1(x)` for some `x`, return that `x`." And in this case, `3` would match `sub1(2)`, so i'd expect it to return `2`?

Also, i'm confused why `x+1` is called `sub1` - i'd have called it `add1`?

Comment (rai-nhdaly) [reply test/rematch.jl]: Oops sorry i left this comment on an outdated commit. Lemme copy it to the latest commit.

File: test/rematch.jl [hunk 7]
```diff
@@ -18,32 +18,248 @@ assertion_error = (VERSION >= v"0.7.0-DEV") ? LoadError : AssertionError
...
-    # variables not bound if match fails
+@testset "Match using extractors" begin
+    function sub1(x) x+1 end
+
+    # sub1(4) == 3
     let x = nothing
-        @test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
+        @test (@match 3 begin
+           ~sub1(x) => x
+        end) == 4
```
Comment (rai-nhdaly) [inline test/rematch.jl:59]: I don't understand why this returns `4`. I could understand if it returned `2` or `3`, but i don't quite understand how it returns `4`.

The definition of `sub1(x)` is `sub1(x) = x+1`. So I take `@match 3 begin   sub1(x) => x  end` to mean "if `3` matches `sub1(x)` for some `x`, return that `x`." And in this case, `3` would match `sub1(2)`, so i'd expect it to return `2`?

Also, i'm confused why `x+1` is called `sub1` - i'd have called it `add1`?

Comment (rai-nhdaly) [reply test/rematch.jl:59]: Ohhhhhhhh after reading some examples, i now understand that in the above match pattern, the `x` doesn't refer to the input, it refers to the _output_.

So `~sub1(x)` means `x == sub1(3)`, meaning `x == 4`.

This was clearer to me after reading the Cons example in the README.


---------------

Perhaps we can find a better syntax that doesn't look like a function application? This would also help with picking a disambiguating syntax from struct constructors.

What do you think about something like this?:
```julia
@test (@match 3 begin
    sub1(~)(x) => x
end) == 4

@match [1,2,3] begin
    Cons(~)(x, xs) => @assert x == 1 && xs == [2,3]
end
```
Where the `~` represents the object being matched on.

Or instead of the `~`, maybe `Cons(_)(x, xs)`? Or `Cons()(x, xs)`? Or maybe something like `Cons()->(x, xs)`, or some other similar but illegal syntax?

Comment (nystrom) [reply test/rematch.jl:59]: > I don't understand why this returns `4`. I could understand if it returned `2` or `3`, but i don't quite understand how it returns `4`.

The idea is that `3 == sub1(4)`. Don't think of `sub1` as a normal function; it's basically the inverse of a constructor.
Maybe we can change the name to `unapply_sub1` or something like that (in Scala extractor methods are all named `unapply`, the inverse of `apply`). This would avoid overloading issues if there's a unary constructor also.

> Perhaps we can find a better syntax that doesn't look like a function application? This would also help with picking a disambiguating syntax from struct constructors.
> 
> What do you think about something like this?:
> 
> ```julia
> @test (@match 3 begin
>     sub1(~)(x) => x
> end) == 4
> 
> @match [1,2,3] begin
>     Cons(~)(x, xs) => @assert x == 1 && xs == [2,3]
> end
> ```

I don't see this as any better than `~Cons(x,xs)`.

Comment (nystrom) [reply test/rematch.jl:59]: Ah, but renaming the function to `unapply_X` breaks parameterized extractors like `Re` in the tests.

Comment (nystrom) [reply test/rematch.jl]: Nevermind, easy to fix. Extractors are named `unapply_Foo` and the pattern is `~Foo`. Pushed.

Comment (nystrom) [reply test/rematch.jl]: After using it a bit, I'm not really sure I like the `unapply_` prefix (or any prefix). Best to just have the writer of the extractor function call it whatever they want, avoiding confusion about which function gets called.

Comment (rai-nhdaly) [reply test/rematch.jl]: > I don't see this as any better than `~Cons(x,xs)`.

The reason I prefer `Cons(~)(x,xs)` is because it prevents parsing the above as a _two argument function call to `Cons`_, which is what it really looks like.

Instead, I want to somehow indicate that the syntax is matching on a two-argument destructured _return value_ from calling `Cons` _on the match variable_. I was imagining the `Cons(~)` syntax to imply "calling `Cons` with the match variable", and then `Cons(~)(x,xs)` implying "the return value from that function can be captured by destructuring into two variables."

Another option might be something like `(x,xs) = Cons(~)`?

Basically, I just want to prevent the initial parse looking like the destructed output variables are actually the inputs to the function, since when they're both arity-1, as in the `sub1` function, it was difficult to understand which variable is the input and which is the output.

Does that make sense? :)

This seems like a really cool feature! I'm sorry to hold it up bikeshedding. Please take this as a positive: this is really cool and I want to help make it intuitive and easy to use! :)

**Comment by amirsh** [comment]
@nystrom this is really nice :)
looking forward to see it merged :D

File: test/rematch.jl (hunk 7 above)
Comment (rai-nhdaly) [reply test/rematch.jl]: Sorry we left this PR lagging for _forever_.

Please feel free to just merge it now, and we can always change this in later iterations if we want to!

After rereading the above discussion, i would like to propose that latest suggestion one more time; i actually think it's fairly nice:

```julia
@test (@match 3 begin
    (x = sub1(~)) => x
end) == 4

@match [1,2,3] begin
    ((x, xs) = Cons(~)) => @assert x == 1 && xs == [2,3]
end
```

This syntax makes the behavior more obvious, IMO:
- call the supplied function on the match variable, and then
- do a capture match on the return value.

----------------------

BUT it seems to me like you are implementing a well known feature in other pattern-matching systems, and I am _not at all_ familiar with those, so please feel free to completely ignore my input. :)

I'd like to get out of the game of causing drama on this repo! ❤️ 😅

**Review by rai-nhdaly — COMMENTED** [review]
@nystrom - sorry for the long delay here. I really don't know much of anything about pattern matching, so it would probably be better to get someone else to review this PR!

@amirsh: Can you maybe review it? 😁 🙏 If you think it's useful, just approve it! We can always make more changes in the future. :) Sorry that i caused such a long hang up!!

**Comment by nystrom** [comment]
Thanks @rai-nhdaly and @amirsh for the comments/review.

I haven't really been pushing this since we've basically worked around its absence for a while now, and I'm no longer sure I'll even use it since extractors will not work with the type dispatch macros I layered on top of `Rematch` in the compiler. I'd like to actually back-port the type dispatch code into `Rematch` but haven't had the time.

Sorry @rai-nhdaly I don't like your syntax proposal :-) It doesn't look much like pattern matching to me anymore. The whole idea is to make the pattern look like the code used to construct the scrutinee of the match, but with free variables for constructor arguments. At least that's what I'm used to in functional languages. I tried to stay as close to the Scala behavior as possible.

I'll merge this when I'm back from vacation.

**Comment by amirsh** [comment]
@NHDaly sure.

Regarding the syntax, I totally agree with @nystrom . The proposed syntax matches better with the one from Scala and F#.

One minor comment is that the last part of the README file is broken due to a missing ` ``` `.

Please let me know once the code is available for review.

**Comment by nystrom** [comment]
I fixed the missing quote, thanks.

**Comment by rai-nhdaly** [comment]
Excellent. Like i said, i don't have much experience with those other systems, so i trust your gut - sorry for the drama. :)

Enjoy the vacation!! :)
